engine/client/gl_rmain.c

```diff
@@ -1326,15 +1327,89 @@ void R_RenderFrame( const ref_params_t *fd, qboolean drawWorld )
 	GL_BackendEndFrame();
 }
 
+inline static void gl_InsertBlackFrame( void )
+{
+	if (CL_IsInConsole()) // No strobing on the console
+	{
+		pglEnable(GL_SCISSOR_TEST);
+		pglScissor(con_rect.x, (-con_rect.y) - (con_rect.h*1.25), con_rect.w, con_rect.h);
+		pglClearColor(0.0f, 0.0f, 0.0f, 1.0f);
+		pglClear(GL_COLOR_BUFFER_BIT);
+		pglDisable(GL_SCISSOR_TEST);
+	}
+	else
+	{
+		pglClearColor(0.0f, 0.0f, 0.0f, 1.0f);
+		pglClear(GL_COLOR_BUFFER_BIT);
+	}
+}
+
+/*
+===============
+R_Strobe
+
+TODO: Consider vsync timings and do not render the supposed black frame at all.
+===============
+*/
+void R_Strobe( void )
+{
+	int getInterval = r_strobe->integer; // Check through modified tag first?
+	int swapInterval = gl_swapInterval->integer;
+	if ( (getInterval == 0) || ((swapInterval == 0) && (getInterval != 0)) )
+	{
+		if (getInterval != 0) //If v-sync is off, turn off strobing
+			Cvar_Set("r_strobe", "0");
+		else if (sCounter != 0)
+			sCounter = 0;
+
+		// flush any remaining 2D bits
+		R_Set2DMode(false);
+		return;
+	}
+	
+	// If interval is positive, insert (replace with) black frames.
+	// For example result of interval = 3 will be: "black-black-black-normal-black-black-black-normal-black-black-black-normal"
+	if (getInterval > 0)
+	{
+		if (sCounter < getInterval)
+		{
+			gl_InsertBlackFrame();
+			++sCounter;
+		}
+		else
+		{
+			sCounter = 0;
+			R_Set2DMode(false);
+		}
+	}
+	// If interval is negative, the procedure will be the opposite reverse.
+	// For example result of interval = -4 will be: "normal-normal-normal-normal-black-normal-normal-normal-normal-black"
+	else
+	{
+		getInterval = abs(getInterval);
+		if (sCounter < getInterval)
+		{
+			++sCounter;
+			R_Set2DMode(false);
+		}
+		else
+		{
+			gl_InsertBlackFrame();
+			sCounter = 0;
+		}
+	}
+}
+
 /*
 ===============
 R_EndFrame
 ===============
 */
 void R_EndFrame( void )
```

Comment:
Before implementing frame skip, it should be tested is frame skip compatible with mod's custom renderers, like Paranoia, HLFX, Trinity or others.

Reply:
I don't know a convenient and efficient way to check if this is compatible with custom renderers. I first thought testing it and checking gl errors but I don't think this is a good way. Maybe it is better to disable this completely for custom renderers (except VBO? ) ?

Reply:
Frame skipping itself may break something even in the engine, so engine side must be completely ready for frame skipping. World VBO is now a part of engine renderer. :)

Yeah, maybe it can be disabled for custom renderers or just filled black at the end of the frame, i.e. no skip.

Reply:
For the current version, I do not think that frame skipping would gain considerable amount of performance increase. So I will not implement it in this pull request and leave that part as is. However I did not check this in any platform other than PC.

My Android build platform is not set. Can you check this in Android and see if it turns out to be bad? Therefore this can be disabled before too late.

Reply:
@fuzun Android version is built automatically by Travis CI. 

There is a link from the end of build log. Script is "travis-upload".
https://transfer.sh/rv8Vy/1024-03-45-xash3d-generic-master-b8bb3c5.apk

Reply:
I can indeed confirm that this is a useful change for 100Hz+ monitors.  It reduces motion blur in the same way as www.testufo.com/blackframes

Reply:
@mdrejhon thanks for professional review! :)

I don't have 60+ Hz screens now, but after reading on your review, it looks like a really cool feature for high frequency screens. So I think it can be merged. :)

Sadly, engine is not ready for frame skipping now and implementing this is a bit complicated because compability with Half-Life, which doesn't have this feature and hardly depends on frametime(which was mostly fixed by Valve in 2013). As it will be done on engine side, maybe someone will use it for BFI.